### ISSUE TITLE
Adaptive capacity management (RateLimiter + BulkHead) decorator

### ISSUE THREAD

**Issue by RobWin** [resilience4j/resilience4j#201]
See concept: https://www.youtube.com/watch?v=m64SWl9bfvk

**Comment by storozhukBM** [comment]
## Implementation thoughts.

As Jon Moore mentioned in his talk we will can increased latency as implicit indication for adaptive concurrency adjustments.

#### Adaptive capacity management prerequisites
1. You should have a system with relatively stable response latency, because we will use latency measures to adapt concurrency limits.
2. To configure you system properly you should figure out two things: 
2.1 Desirable average throughput per second (later **_`X`_**) [Example: 30 req/sec]
2.2 Desirable average request latency in seconds per operation (later **_`R`_**) [Example: 0.1 sec/op]
2.3 Maximum acceptable request latency (later **_`Rmax`_**). This number should be set wisely, because it can eliminate all adaptive capabilities, system will do its best to never reach such latency, so you can set it 20-30 % higher than your usual average latency.

#### Implementation
Resilience4j will provide new Bulkhead implementation called `AdaptiveBulkhead`. It will have following config params:
 1. Desirable average throughput = X
 2. Desirable request latency = R
 3. Maximum acceptable request latency = Rmax (default `R * 1.3`)
 4. Window duration for adaptation = Wa (default 5 sec)
 5. Window duration for reconfiguration = Wr (default 900 sec)
---
From this params we will calculate:
1. Initial average number of concurrent requests (later **_`N`_**).
Example:  `N = X * R` = `30 * 0.1` = `3` **[op]**
2. Initial max latency of current window (later **_`cRmax`_**).
Example:  `cRmax = min(R * 1.2, Rmax)` = `min(0.1 * 1.2, 0.13)` = `0.12` **[op]**
3. Size of adaptation sliding window (later **_`WaN`_**)
Example: `WaN = Wa * X = 30 * 5 = 150`
4. Size of reconfiguration sliding window (later **_`WrN`_**)
Example: `WrN = Wr / Wa = 900 / 5 = 180`
---

From now on we will have two separate functions working inside `AdaptiveBulkhead`:
##### First function is `constantAdaptation`
It will constantly measure request latencies and add them to adaptation sliding window. 
After each cycle with duration of `Wa`, `constantAdaptation` function will calculate current mean latency from adaptation window [**_`cR`_**]. Add this **_`cR`_** to reconfiguration sliding window. Then it will compare **_`cR`_** with **_`cRmax`_**. 
If [**_`cR`_**] < **_`cRmax`_** `constantAdaptation` function will raise number of concurrent requests  **_`N`_** by one, otherwise  it will multiply **_`N`_** by `0.75`.
Also it will use **_`cR`_** to calculate max permission wait time as **_R_** - **_`cR`_**.
 So for next cycle Bulkhead will run with new params.

##### Second function is `configurationAdjustment`
This function will be triggered once per `Wr` and it will calculate standard deviation of latencies from reconfiguration sliding window and will try to recalculate new **_`cRmax`_**.
**_`cRmax`_** = min(**_`cR`_** + standard deviation of reconfiguration window, Rmax)`

This `configurationAdjustment` will help us handle daily latency changes gracefully without reaching **_`cRmax`_** often.

**Comment by storozhukBM** [comment]
@RobWin if you don't have any concerns I'll proceed with implementation.

**Comment by RobWin** [comment]
Can we implement this with a streaming library like reactor? 
We consume events from our existing RateLimiter or Bulkhead and reconfigure them at runtime?

**Comment by storozhukBM** [comment]
Yes, I think it is possible, but such implementation will have some performance drawbacks and will create runtime dependency from streaming library.

**Comment by RobWin** [comment]
I'm looking forward to your implementation :+1:

**Comment by storozhukBM** [comment]
Graphs from the first integration test
## First
![measurement_01](https://user-images.githubusercontent.com/3532750/37991405-bb11709a-3211-11e8-8a93-5773a5079cc4.png)
---
## Second
![measurements_04](https://user-images.githubusercontent.com/3532750/37991430-cb1be48e-3211-11e8-8c19-607f852d8e8e.png)

The main idea is working. Now I'll polish the implementation and make additional performance and integration tests.

**Comment by RobWin** [comment]
Can we progress with this?

**Comment by RobWin** [comment]
I think the multiplicative-decrease factor and additive-increase value should also be configurable.

**Comment by RobWin** [comment]
AIMD requires a signal of congestion. Most frequently, response time outs or bad response times serves as the signal. 
Do you think it also makes to check the content of successful responses if the server signals that the client should back off?

**Comment by storozhukBM** [comment]
Yep, we can provide external API to trigger back off, so our users can use any other type of analysis pluggable or written by hand.

**Comment by liangGTY** [comment]
hi, @RobWin  I would like to ask if there is any progress, when will it be released?

**Comment by RobWin** [comment]
Yes, there is progress. But not sure when it's ready to be released.

**Comment by RobWin** [comment]
![image](https://user-images.githubusercontent.com/5815046/95195826-5e83ef80-07d7-11eb-98b0-2126864c45ea.png)


Our AdaptiveBulkhead implements the TCP Congestion Avoidance algorithm in a protocol-agnostic way.
Instead of a Congestion Window Size, we have a Concurrency Limit.

TCP uses a mechanism called slow start to increase the congestion window after a connection is initialized or after a timeout. It starts with a certain window size. Although the initial rate is low, the rate of increase is very rapid; for every packet acknowledged, the congestion window increases by 1 MSS so that the congestion window effectively doubles for every round-trip time (RTT).

Our AdaptiveBulkhead is a state machine with two states: SLOW_START and CONGESTION_AVOIDANCE.
The user can configure:

Initial Concurrency Limit
Slow Start Concurrency Threshold
Minimum Concurrency Limit
Maximum Concurrency Limit
Increase Summand (default: 1)
Increase Multiplier (default: 2)
Decrease Multiplier (default: 0.5)
In the SLOW_START state the AdaptiveBulkhead will increase the concurrency limit for every successfuly call exponentially (by using "Increase Multiplier") until the slow-start threshold is reached, which is the "Slow Start Concurrency Threshold", then it enters the CONGESTION_AVOIDANCE state. In the CONGESTION_AVOIDANCE state the concurrency limit is increased linearly (by using "Increase Summand").

The AdaptiveBulkhead uses the same Sliding Window implementation like the CircuitBreaker to track successful/failed calls. The Sliding Window is used to calculate a failure rate (and slow call rate) based on the last N calls or calls of the last N seconds. The failure rate increases when certain exceptions occur, e.g. TimeoutException, IOException or TooManyCallsExceptions. If the failure rate (or slow call rate) threshold is exceeded, the AdaptiveBulkhead will multiplicative-decrease the concurrency limit and reset the Sliding Window.

If the Minimum Concurrency Limit is reached, it will enter SLOW_START state again.

Now we need to map this to the AdaptiveRateLimiter ;)
You are right, the AdaptiveRateLimiter shoud not multiplicative-decrease the rate limit on every failed call. But the RateLimiter is currently not tracking a failure rate. What is your suggestion without chaning the existing RateLimiter too much?

**Comment by mehwhatever** [comment]
Hi are we also planning to cover circuit breaker closing using AIMD strategy here? The circuit breaker would gradually move from OPEN -> HALF_OPEN -> CLOSED based on AIMD.

**Comment by RobWin** [comment]
Not sure if an adaptive CircuitBreaker is really nessecarry, if there is an AdaptiveBulkhead based on AIMD. I have the feeling that the changes required in the CircuitBreaker would be too many.

We could rather need support to implement the AdaptiveBulkhead. 
OPEN -> Concurrency Limit = 0
HALF_OPEN is SLOW_START phase
CLSOED is CONGESTION_AVOIDANCE  phase

**Comment by mehwhatever** [comment]
Makes sense. For Adaptive Bulkhead, is the implementation done or can I help in building it?

**Comment by jimhooker2002** [comment]
Also interested in this.  Looks like the implementation is pretty much there now....Some minor points around aesthetics (formatting/spelling) but apart from that looks like it's pretty much there.  Wondering what else would need doing before this would get merged to master and released?

**Comment by RobWin** [comment]
No, it's not done.
We are in the middle of the implementation process -> https://github.com/resilience4j/resilience4j/blob/adaptive_bulkhead_2/resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/internal/AdaptiveBulkheadStateMachine.java

**Comment by RobWin** [comment]
We dismissed the branch `adaptive_bulkhead`

**Comment by jimhooker2002** [comment]
Ah right, you're working on the slow start vs congestion avoidance piece.  Gotcha, thanks for the insight!

**Comment by RoySunnySean007** [comment]
Dear @RobWin ,

Is adaptive_bulkhead_2 similar with https://github.com/Netflix/concurrency-limits?

Thanks,
Roy

**Comment by RobWin** [comment]
I don't know concurrency-limits very well. I can't tell you the differences.
But since our AdaptiveBulkhead is not production-ready, there are not many alternatives to concurrency-limits

**Comment by RoySunnySean007** [comment]
Thanks @RobWin , do we have any ETA to make our AdaptiveBulkhead production-ready? It will be really awesome feature~

**Comment by anhlei** [comment]
Hi, are there any updates on the feature. We're looking to migrate off Hystrix due to the flapping behavior of the bimodal circuit breaker and this feature would be really great

**Comment by RobWin** [comment]
No, unfortunately not.
We are open to any help.

**Comment by RoySunnySean007** [comment]
Dear @RobWin ,

Shameless ask: 
If possible, could u please let us know what is the gap to make adaptive_bulkhead_2 production ready? Could u pls kindly share the sign off criteria?

This feature is really great to us.

Thanks,
Roy

**Comment by RobWin** [comment]
Nobody is currently working on it.
I think @hexmind wants to spend some more time on it.

**Comment by RoySunnySean007** [comment]
Thanks @RobWin !
Dear @hexmind, is it possible to provide answer to my shameless ask https://github.com/resilience4j/resilience4j/issues/201#issuecomment-1370459333 ? 

Thanks in advance!

**Comment by RoySunnySean007** [comment]
That's great! Thanks @hexmind !

**Comment by RoySunnySean007** [comment]
Dear @hexmind , 

I found we have adaptive_bulkhead_3 branch, is it possible to let us know when we will officially release adaptive bulkhead feature? 

Really appreciate ur great help!

**Comment by hexmind** [comment]
Hi, as you see its development has been resumed but it isn't possible to estimate the release date.

**Comment by scott-deboy** [comment]
Hey folks - this could really be useful for a project I'm involved with at work. 

Could someone share what work is outstanding in order to consider this complete and get it merged?

**Comment by rchinmay** [comment]
Hello guys, this seems to be in progress since a very long time. I can join in and help if necessary. Can someone please point out to the current branch where progress is being made, since there seems to be a lot of branches related to this feature?

**Comment by SoftwareApe** [comment]
Is this still actively being worked on? Adaptive Bulkhead is the only thing I'm needing `concurrency-limits` for but I'd prefer to only have one dependency for this.

Is there any work packages that could contribute here?
